### PR TITLE
Allow amend commits to also add --reset-author

### PR DESCRIPTION
**Pull request by joergpichler**
## Proposed changes

- Allows to add --reset-author to an amend commit

### Before

Amending commits with --reset-author was not possible (to my knowledge)

### After

Obviously this PR forces --reset-author for each amend commit. I would like to take this as a basis if this should be in the product at all and if yes, how it should be realized in the UI. UI part has not been done or thought about so far.

## Test methodology 

- Manual testing :)

## Test environment(s) 

- GIT 2.34.1
- Windows 10 21H1

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/FormCommit.cs
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormCommit.cs:1363]: I do not want to change this by default. 
GE should follow the Git defaults when it makes sense, to work predictably.
If the change is minor, then the original author should be preserved.

So unfortunately this need a GUI control, probably a checkbox appearing when amend is checked.

**Review by gerhardol — CHANGES_REQUESTED** [review]

File: GitUI/CommandsDialogs/FormCommit.cs
Comment (joergpichler) [reply GitUI/CommandsDialogs/FormCommit.cs]: I have added an additional checkbox. The enabled state of this checkbox is toggled with the checked state of amend.

grafik

**Comment by vbjay** [comment]
Make sure it is unchecked if amend is unchecked from being checked.

**Comment by joergpichler** [comment]
> Make sure it is unchecked if amend is unchecked from being checked.

Sure. I also cleaned up the UI margins to make all of them equal spaced.

File: GitUI/CommandsDialogs/FormCommit.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs:3266]: I would not show it at all to unclutter the UI.

**Comment by RussKie** [comment]
https://github.com/gitextensions/GitExtensionsDoc will also need to be updated to explain the feature.

File: GitUI/CommandsDialogs/FormCommit.cs
Comment (joergpichler) [reply GitUI/CommandsDialogs/FormCommit.cs:3266]: Will change it.

**Comment by joergpichler** [comment]
> https://github.com/gitextensions/GitExtensionsDoc will also need to be updated to explain the feature.

What is the workflow here? Wait until this is merged and then create a PR in the doc repo?

**Review by gerhardol — APPROVED** [review]
Approving, but do not like how the Reset options jumps when ticking amend.
amend

The core reason is that the Reset buttons are not bottom aligned.
This is likely a legacy from Mono compatibility that did not support such layouts well.
Not really this PR, but since the doc should be updated anyway, it makes sense to change this layout now.

**Comment by gerhardol** [comment]
> What is the workflow here? Wait until this is merged and then create a PR in the doc repo?

Yes (as the PR is approved, you can start now).

The PR build fails as the translation must be updated, required before merging.

**Comment by RussKie** [comment]
I just remembered (apologies for not mentioning this earlier) - the FormCommit has the Options menu, where this functionality _may probably_ fit better, and should address this comment:
> Approving, but do not like how the Reset options jumps when ticking amend.

image

**Comment by joergpichler** [comment]
> I just remembered (apologies for not mentioning this earlier) - the FormCommit has the Options menu, where this functionality _may probably_ fit better, and should address this comment:
> 
> > Approving, but do not like how the Reset options jumps when ticking amend.
> 
> image

Check out my latest changes:
grafik

**Review by gerhardol — APPROVED** [review]
Build still fails due to translation
https://github.com/gitextensions/gitextensions/wiki/Translations#net-50-v36

> > I just remembered (apologies for not mentioning this earlier) - the FormCommit has the Options menu, where this functionality _may probably_ fit better, and should address this comment:
> > > Approving, but do not like how the Reset options jumps when ticking amend.

To some extent I feel that this is a worse solution, this setting is something I expect to be used on case by case basis.
If you set this in settings, then you may have to unset it next time.

But I am OK with the change, I may forget about the option and reset the branch and make a new commit when I occasionally need reset the author.

**Comment by vbjay** [comment]
It would not be set next time. He's just talking about ui placement. You
might as well not have the functionality. It's not very discoverable that
way. Near amend check box makes sense and tying to amend checked makes it
clear. I'd put it in a panel or some other container that would keep the
height there for the other  controls. That way we prevent the bouncing.

Also I'd fix the code I mentioned in my review to help prevent
future issues if commit command is called in other places.

**Comment by joergpichler** [comment]
> It would not be set next time. He's just talking about ui placement. You might as well not have the functionality. It's not very discoverable that way. Near amend check box makes sense and tying to amend checked makes it clear. I'd put it in a panel or some other container that would keep the height there for the other controls. That way we prevent the bouncing. 

I quickly implemented it this way and tbh from all 3 solutions it's the one I personally prefer the most. I am not sure if I should update the PR again with this change before it's clear which solution will be picked in the end.

> Also I'd fix the code I mentioned in my review to help prevent future issues if commit command is called in other places.

I don't really get that point, can you elaborate further?

**Comment by vbjay** [comment]
> > Also I'd fix the code I mentioned in my review to help prevent future issues if commit command is called in other places.
> 
> I don't really get that point, can you elaborate further?
> 

https://github.com/gitextensions/gitextensions/pull/9778#pullrequestreview-842130984

**Comment by joergpichler** [comment]
I have updated the PR to use the solution that is probably the most preferred here: Having the checkbox next to the Amend checkbox but only visible when Amend is checked without the UI jumping
GitExtensions_9s0wHkj2ju

**Review by gerhardol — APPROVED** [review]

**Comment by vbjay** [comment]
So basically you think my point is wrong or are  unwilling to do 

https://github.com/gitextensions/gitextensions/pull/9778#pullrequestreview-842130984

**Comment by joergpichler** [comment]
> So basically you think my point is wrong or are unwilling to do
> 
> [#9778 (review)](https://github.com/gitextensions/gitextensions/pull/9778#pullrequestreview-842130984)

I'm truly sorry but neither is the case. Clicking on this link goes nowhere for me, so I still don't understand what you are trying to say.

**Comment by pmiossec** [comment]
> So basically you think my point is wrong or are unwilling to do
> 
> [#9778 (review)](https://github.com/gitextensions/gitextensions/pull/9778#pullrequestreview-842130984)

@vbjay link broken also for me so you will have to do your review again or say again what you want to say...

**Comment by vbjay** [comment]
Ok now that makes more sense. Go to changes tab in the pull request and look at my review.

**Review by vbjay — CHANGES_REQUESTED** [review]

**Comment by vbjay** [comment]
Omg. GitHub....It didn't submit it. When I thought I did. No wonder.

**Review by RussKie — COMMENTED** [review]
Please add tests to `FormCommitTests` to verify the expected behaviours.

**Comment by RussKie** [comment]
Also please update the translations. Refer to https://github.com/gitextensions/gitextensions/wiki/Translations

File: GitCommands/Git/GitModule.cs
Comment (joergpichler) [reply GitCommands/Git/GitModule.cs:2147]: Sure makes sense. Will update it

**Comment by joergpichler** [comment]
I have added the test, ran the translation command and included the changes from @vbjay

File: GitUI/CommandsDialogs/FormCommit.Designer.cs
Comment (mstv) [inline GitUI/CommandsDialogs/FormCommit.Designer.cs:1177]: Please avoid conflict with `this.Reset` "&Reset all changes":
```suggestion
            this.ResetAuthor.Text = "R&eset Author";
```

**Review by mstv — COMMENTED** [review]
+1

**Comment by joergpichler** [comment]
> +1

I have updated the PR accordingly

**Comment by mstv** [comment]
Squash merge (the single commit) in three days?

**Pull request merged**
